### ISSUE TITLE
"Cannot read property 'filter' of undefined" when using postcss-less

### ISSUE THREAD

**Issue by trisys3** [ben-eb/cssnano#373]
I am not sure if this is the right place to put this. I know next to nothing about PostCSS internals or the plugin architecture. I am intending to use [postcss-less](https://www.npmjs.com/package/postcss-less) until we can switch to SCSS, mainly because Bootstrap is switching over. I am hoping we can use both for a time, but I am not optimistic.

Basically, combining `cssnano` with `postcss-less` is causing `cssnano` to fail the build if any LESS mixins are called. For some reason, the mixin must be called: defining it allows the build to compile. The same does not happen with [postcss-scss](https://www.npmjs.com/package/postcss-scss), even with mixins. One simple case is:

    /** app.less **/
    .addColor(@color) {
        color: @color;
    }

    body {
        // .addColor(black);
    }

    /** app/app.scss **/
    @mixin addColor($color) {
        background-color: $color;
    }

    body {
        @include addColor(black);
    }

If you uncomment the mixin call in the LESS, the build breaks, but the SCSS works fine. If I try to use the LESS version, uncommented, I get:

```
internal/streams/legacy.js:59
      throw er; // Unhandled stream error in pipe.
      ^
TypeError: Cannot read property 'filter' of undefined
    at /repo/folder/node_modules/postcss-merge-longhand/dist/index.js:71:23
    at /repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:241:28
    at /repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:148:26
    at Rule.each (/repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:114:22)
    at Rule.walk (/repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:147:21)
    at /repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:150:32
    at Root.each (/repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:114:22)
    at Root.walk (/repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:147:21)
    at Root.walkRules (/repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:239:25)
    at /repo/folder/node_modules/postcss-merge-longhand/dist/index.js:70:13
```

As the same thing that happens with `postcss-less` does not happen with `postcss-scss`, I was thinking the issue might be with `postcss-less`, but then I read a note on the `postcss-less` README, under the `rule.nodes` section, saying that `rule.nodes` may be undefined for rules that don't contain declarations. As I said, I know almost nothing about the internals of this plugin, but I remembered that `rule.nodes` was also mentioned in the error message. Is this sort of thing expected? Should I report this issue to `postcss-less` instead, or even `postcss-merge-longhand`?

I am using `gulp` with `gulp-postcss`, but I do not think this matters. The basic structure of the task is:

    const nano = require('cssnano');
    const lessParser = require('postcss-less');
    const scssParser = require('postcss-scss');

    const preprocessor = process.env.CSS_PREPROCESSOR;
    const cssSrc = `app/app.${preprocessor}`;

    let syntax;
    if(preprocessor === 'less') {
        syntax = lessParser;
    } else if(preprocessor === 'scss') {
        syntax = scssParser;
    }

    gulp.src(cssSrc)
        .pipe(postcss({
            plugins: [
                nano(),
            ],
        }, {
            syntax,
        });

Change the environment variable of CSS_PREPROCESSOR from "less" to "scss" to change from parsing LESS to parsing SCSS, and remove the variable to not use a preprocessor.

versions:
  `postcss-less`: 1.0.1
  `postcss-scss`: 1.0.0
  `cssnano`: 3.10.0

**Comment by ben-eb** [comment]
@trisys3 Can you reproduce with [4.0.0-rc.0](https://github.com/ben-eb/cssnano/releases/tag/v4.0.0-rc.0)? `npm i cssnano@next`

**Comment by trisys3** [comment]
OK, hold on. I updated my issue with the versions of everything that I used.

**Comment by trisys3** [comment]
Yes, it still happens. Slightly different error message, though:

```
internal/streams/legacy.js:59
      throw er; // Unhandled stream error in pipe.
      ^
TypeError: Cannot read property 'filter' of undefined
    at getDecls (/repo/folder/node_modules/postcss-merge-longhand/dist/lib/getDecls.js:8:22)
    at mergeRules (/repo/folder/node_modules/postcss-merge-longhand/dist/lib/mergeRules.js:25:40)
    at /repo/folder/node_modules/postcss-merge-longhand/dist/lib/decl/borders.js:211:34
    at Array.forEach (native)
    at Object.merge (/repo/folder/node_modules/postcss-merge-longhand/dist/lib/decl/borders.js:209:20)
    at /repo/folder/node_modules/postcss-merge-longhand/dist/index.js:22:19
    at Array.forEach (native)
    at /repo/folder/node_modules/postcss-merge-longhand/dist/index.js:20:28
    at /repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:241:28
    at /repo/folder/node_modules/postcss-less/node_modules/postcss/lib/container.js:148:26
```

**Comment by ben-eb** [comment]
Remember that cssnano expects to work with CSS and not preprocessor syntax; postcss-less doesn't compile less into CSS. For that you'll want https://www.npmjs.com/package/postcss-less-engine.

**Comment by andyjansson** [comment]
@trisys3 It looks to me like you've misunderstood the purpose of postcss-less. Like it says at the top of the readme: 

> Please note: poscss-less is not a LESS compiler. For compiling LESS, please use the official toolset for LESS.

What it does is allow postcss to read LESS syntax; it does not execute it. By running cssnano with postcss-less, all you're doing is minifying LESS code (or in this case, failing, as cssnano expects valid CSS).

**Comment by trisys3** [comment]
You are right, I was confused by the `postcss-less`/`postcss-scss` documentation. I saw that these are used to parse out the LESS/SCSS, and I understood that the plugins would not compile and I would still need a compile step. I suppose I did not understand that not all plugins would be able to handle the parsed syntax, though that seems obvious now. Thank you for clarifying what should have been abundantly clear already.
